File: keyboards/ios/LisanUdDawat/LSDLearningKB/BiDiFixMenu.swift
import UIKit

// Slide-up panel offering BiDi fix options.
// Shown when the user taps the animated BiDi fix button in the predictive bar.

final class BiDiFixMenu: UIView {

    var onFixApplied: ((BiDiFix) -> Void)?
    var onDismiss:    (() -> Void)?

    enum BiDiFix { case smart, fixLine, fixAtCursor, markSelectionLTR }

    private let issue: BiDiAnalyzer.Issue

    init(issue: BiDiAnalyzer.Issue) {
        self.issue = issue
        super.init(frame: .zero)
        buildUI()
    }
    required init?(coder: NSCoder) { fatalError() }

    // MARK: - Show

    static func show(in parent: UIView, issue: BiDiAnalyzer.Issue) -> BiDiFixMenu {
        let menu = BiDiFixMenu(issue: issue)
        menu.translatesAutoresizingMaskIntoConstraints = false
        parent.addSubview(menu)
        NSLayoutConstraint.activate([
            menu.topAnchor.constraint(equalTo: parent.topAnchor),
            menu.bottomAnchor.constraint(equalTo: parent.bottomAnchor),
            menu.leadingAnchor.constraint(equalTo: parent.leadingAnchor),
            menu.trailingAnchor.constraint(equalTo: parent.trailingAnchor),
        ])
        parent.layoutIfNeeded()
        menu.transform = CGAffineTransform(translationX: 0, y: parent.bounds.height)
        UIView.animate(withDuration: 0.22, delay: 0, options: .curveEaseOut) {
            menu.transform = .identity
        }
        return menu
    }

    func dismiss() {
        UIView.animate(withDuration: 0.18, delay: 0, options: .curveEaseIn) {
            self.transform = CGAffineTransform(translationX: 0, y: self.bounds.height)
        } completion: { _ in
            self.removeFromSuperview()
            self.onDismiss?()
        }
    }

    // MARK: - Build

    private func buildUI() {
        // Tapping the backdrop dismisses
        isUserInteractionEnabled = true
        addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(backdropTapped)))
        backgroundColor = .clear

        let panel = UIView()
        panel.backgroundColor = KeyboardColors.background
        panel.translatesAutoresizingMaskIntoConstraints = false
        addSubview(panel)
        NSLayoutConstraint.activate([
            panel.leadingAnchor.constraint(equalTo: leadingAnchor),
            panel.trailingAnchor.constraint(equalTo: trailingAnchor),
            panel.bottomAnchor.constraint(equalTo: bottomAnchor),
        ])
        // Absorb taps so they don't pass through to the backdrop
        panel.addGestureRecognizer(UITapGestureRecognizer(target: nil, action: nil))

        let header = buildHeader()
        panel.addSubview(header)
        NSLayoutConstraint.activate([
            header.topAnchor.constraint(equalTo: panel.topAnchor),
            header.leadingAnchor.constraint(equalTo: panel.leadingAnchor),
            header.trailingAnchor.constraint(equalTo: panel.trailingAnchor),
            header.heightAnchor.constraint(equalToConstant: 44),
        ])

        let sep = hairline()
        panel.addSubview(sep)
        NSLayoutConstraint.activate([
            sep.topAnchor.constraint(equalTo: header.bottomAnchor),
            sep.leadingAnchor.constraint(equalTo: panel.leadingAnchor),
            sep.trailingAnchor.constraint(equalTo: panel.trailingAnchor),
        ])

        let stack = UIStackView()
        stack.axis = .vertical
        stack.spacing = 0
        stack.translatesAutoresizingMaskIntoConstraints = false
        panel.addSubview(stack)
        NSLayoutConstraint.activate([
            stack.topAnchor.constraint(equalTo: sep.bottomAnchor, constant: 8),
            stack.leadingAnchor.constraint(equalTo: panel.leadingAnchor, constant: 16),
            stack.trailingAnchor.constraint(equalTo: panel.trailingAnchor, constant: -16),
            stack.bottomAnchor.constraint(equalTo: panel.bottomAnchor, constant: -16),
        ])

        // Primary "Fix" button
        let fixBtn = UIButton(type: .system)
        fixBtn.setTitle("Fix", for: .normal)
        fixBtn.titleLabel?.font = .systemFont(ofSize: 16, weight: .bold)
        fixBtn.backgroundColor = UIColor(red: 0, green: 0.478, blue: 1, alpha: 1)
        fixBtn.setTitleColor(.white, for: .normal)
        fixBtn.layer.cornerRadius = 10
        fixBtn.translatesAutoresizingMaskIntoConstraints = false
        fixBtn.heightAnchor.constraint(equalToConstant: 48).isActive = true
        fixBtn.addTarget(self, action: #selector(fixTapped), for: .touchUpInside)
        stack.addArrangedSubview(fixBtn)

        let spacer = UIView()
        spacer.translatesAutoresizingMaskIntoConstraints = false
        spacer.heightAnchor.constraint(equalToConstant: 10).isActive = true
        stack.addArrangedSubview(spacer)

        stack.addArrangedSubview(hairline())
        addRow(to: stack,
               title: "Fix this line",
               subtitle: "Adds a direction mark at the start of this line",
               action: { [weak self] in self?.applyFix(.fixLine) })
        stack.addArrangedSubview(hairline())
        addRow(to: stack,
               title: "Fix at cursor",
               subtitle: "Adds a direction mark at the cursor position",
               action: { [weak self] in self?.applyFix(.fixAtCursor) })
        stack.addArrangedSubview(hairline())
        addRow(to: stack,
               title: "Mark selection as left-to-right",
               subtitle: "Wraps the selected text so it reads left-to-right",
               action: { [weak self] in self?.applyFix(.markSelectionLTR) })
    }

    private func buildHeader() -> UIView {
        let header = UIView()
        header.backgroundColor = KeyboardColors.predictiveBar
        header.translatesAutoresizingMaskIntoConstraints = false

        let title = UILabel()
        title.text = "Fix Text Direction"
        title.font = .systemFont(ofSize: 14, weight: .semibold)
        title.textColor = .label
        title.translatesAutoresizingMaskIntoConstraints = false
        header.addSubview(title)

        let done = UIButton(type: .system)
        done.setTitle("Done", for: .normal)
        done.titleLabel?.font = .systemFont(ofSize: 14, weight: .semibold)
        done.translatesAutoresizingMaskIntoConstraints = false
        done.addTarget(self, action: #selector(backdropTapped), for: .touchUpInside)
        header.addSubview(done)

        NSLayoutConstraint.activate([
            title.centerXAnchor.constraint(equalTo: header.centerXAnchor),
            title.centerYAnchor.constraint(equalTo: header.centerYAnchor),
            done.trailingAnchor.constraint(equalTo: header.trailingAnchor, constant: -14),
            done.centerYAnchor.constraint(equalTo: header.centerYAnchor),
        ])
        return header
    }

    private func addRow(to stack: UIStackView, title: String, subtitle: String, action: @escaping () -> Void) {
        let row = UIControl()
        row.translatesAutoresizingMaskIntoConstraints = false
        row.heightAnchor.constraint(equalToConstant: 52).isActive = true
        row.addAction(UIAction { _ in action() }, for: .touchUpInside)

        let vstack = UIStackView()
        vstack.axis = .vertical
        vstack.spacing = 2
        vstack.isUserInteractionEnabled = false
        vstack.translatesAutoresizingMaskIntoConstraints = false

        let titleLbl = UILabel()
        titleLbl.text = title
        titleLbl.font = .systemFont(ofSize: 14)
        titleLbl.textColor = .label

        let subLbl = UILabel()
        subLbl.text = subtitle
        subLbl.font = .systemFont(ofSize: 11)
        subLbl.textColor = .secondaryLabel
        subLbl.numberOfLines = 2

        vstack.addArrangedSubview(titleLbl)
        vstack.addArrangedSubview(subLbl)
        row.addSubview(vstack)
        NSLayoutConstraint.activate([
            vstack.leadingAnchor.constraint(equalTo: row.leadingAnchor),
            vstack.trailingAnchor.constraint(equalTo: row.trailingAnchor),
            vstack.centerYAnchor.constraint(equalTo: row.centerYAnchor),
        ])
        stack.addArrangedSubview(row)
    }

    private func hairline() -> UIView {
        let v = UIView()
        v.backgroundColor = KeyboardColors.separator
        v.translatesAutoresizingMaskIntoConstraints = false
        v.heightAnchor.constraint(equalToConstant: 0.5).isActive = true
        return v
    }

    // MARK: - Actions

    @objc private func fixTapped()      { applyFix(.smart) }
    @objc private func backdropTapped() { dismiss() }

    private func applyFix(_ fix: BiDiFix) {
        onFixApplied?(fix)
        dismiss()
    }
}

File: keyboards/android/app/src/main/java/com/exordiumnetworks/ligacheh/BiDiFixMenu.kt
package com.exordiumnetworks.ligacheh

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.view.Gravity
import android.view.View
import android.view.inputmethod.InputConnection
import android.widget.*

// Popup panel shown when the BiDi fix button is tapped.
// Offers a smart one-tap Fix plus three manual options.
class BiDiFixMenu(context: Context) : FrameLayout(context) {

    companion object {
        fun show(
            parent: FrameLayout,
            issue: BiDiAnalyzer.Issue?,
            getIc: () -> InputConnection?,
            onFixApplied: () -> Unit,
            onDismiss: () -> Unit
        ): BiDiFixMenu {
            val menu = BiDiFixMenu(parent.context)
            menu.currentIssue  = issue
            menu.getIc         = getIc
            menu.onFixApplied  = onFixApplied
            menu.onDismiss     = onDismiss
            parent.addView(menu, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
            val startY = parent.height.toFloat().coerceAtLeast(300f)
            menu.translationY = startY
            parent.post { menu.animate().translationY(0f).setDuration(200).start() }
            return menu
        }
    }

    private var currentIssue: BiDiAnalyzer.Issue? = null
    private var getIc: (() -> InputConnection?)? = null
    private var onFixApplied: (() -> Unit)? = null
    private var onDismiss: (() -> Unit)? = null

    init { buildUi() }

    private fun applyAndDismiss(action: (InputConnection) -> Unit) {
        val ic = getIc?.invoke() ?: return
        action(ic)
        onFixApplied?.invoke()
        dismiss()
    }

    private fun buildUi() {
        setOnClickListener { dismiss() }
        setBackgroundColor(Color.TRANSPARENT)

        val panel = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            setBackgroundColor(KeyboardColors.background(context))
            setOnClickListener { /* consume */ }
        }
        val panelLp = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        panelLp.gravity = Gravity.BOTTOM
        addView(panel, panelLp)

        panel.addView(buildHeader())
        panel.addView(hairline())

        val content = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(16), dp(8), dp(16), dp(16))
        }
        panel.addView(content)

        // Smart Fix button
        val fixBtn = TextView(context).apply {
            text      = "Fix"
            textSize  = 16f
            setTypeface(null, Typeface.BOLD)
            setTextColor(Color.WHITE)
            gravity   = Gravity.CENTER
            background = GradientDrawable().apply {
                setColor(Color.parseColor("#007AFF"))
                cornerRadius = dp(10f)
            }
            isClickable = true; isFocusable = true
            setOnClickListener {
                applyAndDismiss { ic -> currentIssue?.let { BiDiAnalyzer.applySmartFix(it, ic) } }
            }
        }
        val fixLp = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, dp(48))
        fixLp.bottomMargin = dp(10)
        content.addView(fixBtn, fixLp)

        content.addView(hairline())
        addManualRow(content,
            "Fix this line",
            "Adds a direction mark at the start of this line") {
            applyAndDismiss { BiDiAnalyzer.insertRlmAtLineStart(it) }
        }
        content.addView(hairline())
        addManualRow(content,
            "Fix at cursor",
            "Adds a direction mark at the cursor position") {
            applyAndDismiss { BiDiAnalyzer.insertRlmAtCursor(it) }
        }
        content.addView(hairline())
        addManualRow(content,
            "Mark selection as left-to-right",
            "Wraps the selected text so it reads left-to-right") {
            applyAndDismiss { BiDiAnalyzer.wrapSelectionAsLtr(it) }
        }
    }

    private fun buildHeader(): View {
        val bar = FrameLayout(context).apply {
            setBackgroundColor(KeyboardColors.predictiveBar(context))
        }
        bar.addView(TextView(context).apply {
            text      = "Fix Text Direction"
            textSize  = 14f
            setTypeface(null, Typeface.BOLD)
            setTextColor(KeyboardColors.keyLabel(context))
            gravity   = Gravity.CENTER
        }, FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, dp(44)))
        bar.addView(TextView(context).apply {
            text      = "Done"
            textSize  = 14f
            setTypeface(null, Typeface.BOLD)
            setTextColor(Color.parseColor("#007AFF"))
            setPadding(0, 0, dp(14), 0)
            gravity   = Gravity.CENTER_VERTICAL or Gravity.END
            isClickable = true; isFocusable = true
            setOnClickListener { dismiss() }
        }, FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, dp(44)))
        return bar
    }

    private fun addManualRow(parent: LinearLayout, label: String, sub: String, action: () -> Unit) {
        val row = LinearLayout(context).apply {
            orientation  = LinearLayout.VERTICAL
            gravity      = Gravity.CENTER_VERTICAL
            layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, dp(52))
            isClickable  = true; isFocusable = true
            setOnClickListener { action() }
        }
        row.addView(TextView(context).apply {
            text     = label
            textSize = 14f
            setTextColor(KeyboardColors.keyLabel(context))
        })
        row.addView(TextView(context).apply {
            text     = sub
            textSize = 11f
            val base = KeyboardColors.keyLabel(context)
            setTextColor(Color.argb(140, Color.red(base), Color.green(base), Color.blue(base)))
        })
        parent.addView(row)
    }

    fun dismiss() {
        animate().translationY(height.toFloat()).setDuration(160).withEndAction {
            (parent as? android.view.ViewGroup)?.removeView(this)
            onDismiss?.invoke()
        }.start()
    }

    private fun hairline() = View(context).apply {
        setBackgroundColor(KeyboardColors.separator(context))
        layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1)
    }

    private fun dp(v: Int)   = (v * resources.displayMetrics.density).toInt()
    private fun dp(v: Float) = v * resources.displayMetrics.density
}
